File: Sean的leetcode/Leetcode/Super Ugly Number.cpp
/*
 * smallidx: a prime in primes can multiply an element in ugly with the smallest index
 * Time Complexity: O(nk)
 */
class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {
        vector<int> smallidx(primes.size(), 0);
        vector<int> ugly(n, INT_MAX);
        ugly[0] = 1;
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < primes.size(); j++) {
                ugly[i] = min(ugly[i], ugly[smallidx[j]] * primes[j]);
            }
            for (int j = 0; j < primes.size(); j++) {
                smallidx[j] += (ugly[i] == ugly[smallidx[j]] * primes[j]);
            }
        }
        return ugly[n - 1];
    }
};

File: MyLeetCode/python/Super Ugly Number.py
###
# Priority Queue
# Time Complexity: O(k) + O(nlogk)
# Space Complexity: O(k)
###
class Solution(object):
    def nthSuperUglyNumber(self, n, primes):
        """
        :type n: int
        :type primes: List[int]
        :rtype: int
        """
        import heapq
        length = len(primes)
        small_idx = [0] * length
        ugly = [1]
        q  = [(primes[i] * ugly[small_idx[i]], i) for i in xrange(len(primes))]
        heapq.heapify(q)
        while len(ugly) < n:
            tmp, idx = heapq.heappop(q)
            small_idx[idx] += 1
            if tmp != ugly[-1]:
                ugly.append(tmp)
            heapq.heappush(q, (primes[idx] * ugly[small_idx[idx]], idx))
        return ugly[-1] 
        
        